File: tkt_md/tickets/repair.py
from __future__ import annotations
from pathlib import Path
from typing import List, Dict, Any, Tuple, Optional

import yaml

from . import util, validation


def load_issues_file(path: Path) -> Dict[str, Any]:
    return yaml.safe_load(path.read_text()) or {}


def apply_repairs(repairs: List[Dict[str, Any]], non_interactive: bool = False, include_optional: bool = False) -> List[str]:
    applied = []
    for rep in repairs:
        if rep.get("optional") and not include_optional:
            continue
        if not rep.get("enabled"):
            continue
        action = rep.get("action")
        params = rep.get("params", {})
        ticket_path = Path(rep.get("ticket_path", ""))

        if action == "set_front_matter_field":
            field = params.get("field")
            value = params.get("value")
            if value is None and params.get("generate_uuidv7"):
                value = util.new_uuidv7()
            if value is None:
                if non_interactive:
                    raise RuntimeError(f"Repair needs value for {field}")
                value = input(f"Value for {field}: ").strip()
            _set_front_matter_field(ticket_path, field, value)
            applied.append(f"{ticket_path}: set {field}")

        elif action == "add_sections":
            _add_missing_sections(ticket_path)
            applied.append(f"{ticket_path}: added missing sections")

        elif action == "normalize_created_at":
            _normalize_created_at(ticket_path)
            applied.append(f"{ticket_path}: normalized created_at")

        elif action == "normalize_labels":
            _normalize_labels(ticket_path)
            applied.append(f"{ticket_path}: normalized labels")

        elif action == "set_assignment_owner":
            _set_assignment_owner(ticket_path, params.get("value"))
            applied.append(f"{ticket_path}: set assignment.owner")

        elif action == "reset_verification_commands":
            _reset_verification_commands(ticket_path, params.get("commands", []))
            applied.append(f"{ticket_path}: reset verification.commands")

        elif action == "normalize_verification_commands":
            _normalize_verification_commands(ticket_path)
            applied.append(f"{ticket_path}: normalized verification.commands")

        else:
            if non_interactive:
                raise RuntimeError(f"Unsupported repair action {action}")
    return applied


def run_interactive(repairs: List[Dict[str, Any]], include_optional: bool = False) -> List[str]:
    """
    Step through repairs, explain them, allow user to confirm/apply and supply values.
    Returns list of applied changes (same shape as apply_repairs).
    """
    prepared = []
    for rep in repairs:
        if rep.get("optional") and not include_optional:
            continue
        desc, default_value = _describe_repair(rep)
        print(f"\nRepair {rep.get('id')}: {desc}")
        apply_choice = _prompt_yes_no("Apply this repair?", default=True)
        if not apply_choice:
            continue
        # enable and fill values if needed
        rep["enabled"] = True
        action = rep.get("action")
        params = rep.get("params", {})
        if action == "set_front_matter_field":
            field = params.get("field")
            suggested = default_value
            value = _prompt_value_for_field(field, rep.get("ticket_path"), suggested)
            params["value"] = value
            rep["params"] = params
        elif action == "set_assignment_owner":
            value = _prompt_value_for_field("assignment.owner", rep.get("ticket_path"), default_value)
            params["value"] = value
            rep["params"] = params
        elif action == "reset_verification_commands":
            value = _prompt_commands(default_value or [])
            params["commands"] = value
            rep["params"] = params
        elif action == "normalize_verification_commands":
            # no extra input, but keep enabled
            pass
        prepared.append(rep)
    # apply with non_interactive=True because choices/values already captured
    return apply_repairs(prepared, non_interactive=True, include_optional=include_optional)


def _describe_repair(rep: Dict[str, Any]) -> Tuple[str, Optional[Any]]:
    action = rep.get("action")
    field = (rep.get("params") or {}).get("field")
    path = rep.get("ticket_path", "")
    default = (rep.get("params") or {}).get("value")
    if action == "add_sections":
        return (f"Add missing required sections to {path}.", None)
    if action == "normalize_created_at":
        return (f"Normalize created_at to ISO8601 UTC in {path}.", util.iso8601(util.now_utc()))
    if action == "set_front_matter_field":
        if field == "id":
            return ("Set ticket id to a valid UUIDv7 (used to identify the ticket).", util.new_uuidv7())
        if field == "version":
            return ("Set format version (integer, current 1).", util.FORMAT_VERSION)
        if field == "version_url":
            return ("Set version_url (path to the format definition for this version).", util.FORMAT_VERSION_URL)
        if field == "priority":
            return ("Set priority (low|medium|high|critical).", "medium")
        if field == "labels":
            return ("Reset labels to a list of strings (comma-separated).", [])
        return (f"Set front matter field '{field}'.", default)
    if action == "normalize_labels":
        return ("Normalize labels to strings, dropping invalid entries.", None)
    if action == "set_assignment_owner":
        return ("Set assignment.owner (who owns this ticket; freeform handle).", default)
    if action == "reset_verification_commands":
        return ("Set verification.commands (commands to verify acceptance).", default or [])
    if action == "normalize_verification_commands":
        return ("Normalize verification.commands to strings, dropping invalid entries.", None)
    return ("Apply repair", default)


def _prompt_yes_no(message: str, default: bool = True) -> bool:
    prompt = " [Y/n] " if default else " [y/N] "
    while True:
        resp = input(message + prompt).strip().lower()
        if not resp:
            return default
        if resp in ["y", "yes"]:
            return True
        if resp in ["n", "no"]:
            return False
        print("Please enter y or n.")


def _prompt_value_for_field(field: str, ticket_path: str | None, default: Any) -> Any:
    current = None
    if ticket_path:
        try:
            fm, _ = util.load_ticket(Path(ticket_path))
            if field == "assignment.owner":
                current = (fm.get("assignment") or {}).get("owner")
            else:
                current = fm.get(field)
        except Exception:
            current = None
    if field == "labels":
        return _prompt_labels(default if default is not None else (current if current is not None else []))
    if field == "priority":
        choices = ["low", "medium", "high", "critical"]
        default_val = default or current or "medium"
        while True:
            resp = input(f"Priority [{default_val}]: ").strip().lower()
            if not resp:
                return default_val
            if resp in choices:
                return resp
            print(f"Enter one of {choices}.")
    value_default = default if default is not None else current
    resp = input(f"{field} [{value_default if value_default is not None else ''}]: ").strip()
    if resp == "" and value_default is not None:
        return value_default
    if field == "id":
        return resp or util.new_uuidv7()
    return resp if resp != "" else value_default


def _prompt_labels(default_labels: List[str]) -> List[str]:
    existing = ", ".join(default_labels) if default_labels else ""
    resp = input(f"Labels (comma-separated) [{existing}]: ").strip()
    if not resp:
        return default_labels
    return [s.strip() for s in resp.split(",") if s.strip()]


def _prompt_commands(default_commands: List[str]) -> List[str]:
    existing = "; ".join(default_commands) if default_commands else ""
    print("Enter verification commands (comma-separated). Leave blank to keep default.")
    resp = input(f"Commands [{existing}]: ").strip()
    if not resp:
        return default_commands
    return [s.strip() for s in resp.split(",") if s.strip()]


def _set_front_matter_field(ticket_path: Path, field: str, value: Any):
    fm, body = util.load_ticket(ticket_path)
    fm[field] = value
    util.write_ticket(ticket_path, fm, body)


def _add_missing_sections(ticket_path: Path):
    fm, body = util.load_ticket(ticket_path)
    sections = ["# Ticket", "## Description", "## Acceptance Criteria", "## Verification"]
    new_body = body
    for sec in sections:
        if sec not in new_body:
            new_body += f"\n{sec}\n(fill in)\n"
    util.write_ticket(ticket_path, fm, new_body)


def _normalize_created_at(ticket_path: Path):
    fm, body = util.load_ticket(ticket_path)
    val = fm.get("created_at")
    if not isinstance(val, str) or util.parse_iso(val) is None:
        fm["created_at"] = util.iso8601(util.now_utc())
        util.write_ticket(ticket_path, fm, body)


def _normalize_labels(ticket_path: Path):
    fm, body = util.load_ticket(ticket_path)
    labels = fm.get("labels")
    normalized = [v for v in labels if isinstance(v, str)] if isinstance(labels, list) else []
    fm["labels"] = normalized
    util.write_ticket(ticket_path, fm, body)


def _set_assignment_owner(ticket_path: Path, value: Any):
    fm, body = util.load_ticket(ticket_path)
    assignment = fm.get("assignment") if isinstance(fm.get("assignment"), dict) else {}
    assignment["owner"] = value
    fm["assignment"] = assignment
    util.write_ticket(ticket_path, fm, body)


def _reset_verification_commands(ticket_path: Path, commands: List[str]):
    fm, body = util.load_ticket(ticket_path)
    fm["verification"] = {"commands": commands}
    util.write_ticket(ticket_path, fm, body)


def _normalize_verification_commands(ticket_path: Path):
    fm, body = util.load_ticket(ticket_path)
    ver = fm.get("verification") if isinstance(fm.get("verification"), dict) else {}
    cmds_raw = ver.get("commands") if isinstance(ver, dict) else None
    cmds = [c for c in cmds_raw if isinstance(c, str)] if isinstance(cmds_raw, list) else []
    fm["verification"] = {"commands": cmds}
    util.write_ticket(ticket_path, fm, body)

File: packages/tickets/src/lib/repair.js
import fs from "node:fs";
import readline from "node:readline/promises";

import yaml from "yaml";

import {
  FORMAT_VERSION,
  FORMAT_VERSION_URL,
  PRIORITY_VALUES,
} from "./constants.js";
import {
  iso8601,
  loadTicket,
  newUuidv7,
  nowUtc,
  parseIso,
  readJsonl,
  writeTicket,
} from "./util.js";

export function loadIssuesFile(filePath) {
  return yaml.parse(fs.readFileSync(filePath, "utf8")) ?? {};
}

export async function applyRepairs(repairs, options = {}) {
  const includeOptional = options.includeOptional ?? false;
  const nonInteractive = options.nonInteractive ?? false;
  const applied = [];

  for (const repair of repairs) {
    if (repair.optional && !includeOptional) {
      continue;
    }
    if (!repair.enabled) {
      continue;
    }

    const action = repair.action;
    const params = repair.params ?? {};
    const ticketPath = repair.ticket_path;
    const logPath = repair.log_path;

    if (action === "set_front_matter_field") {
      const field = params.field;
      let value = params.value;
      if (value == null && params.generate_uuidv7) {
        value = newUuidv7();
      }
      if (value == null) {
        if (nonInteractive) {
          throw new Error(`Repair needs value for ${field}`);
        }
        throw new Error(`Interactive value required for ${field}`);
      }
      setFrontMatterField(ticketPath, field, value);
      applied.push(`${ticketPath}: set ${field}`);
      continue;
    }

    if (action === "add_sections") {
      addMissingSections(ticketPath);
      applied.push(`${ticketPath}: added missing sections`);
      continue;
    }

    if (action === "normalize_created_at") {
      normalizeCreatedAt(ticketPath);
      applied.push(`${ticketPath}: normalized created_at`);
      continue;
    }

    if (action === "normalize_labels") {
      normalizeLabels(ticketPath);
      applied.push(`${ticketPath}: normalized labels`);
      continue;
    }

    if (action === "set_assignment_owner") {
      setAssignmentOwner(ticketPath, params.value ?? null);
      applied.push(`${ticketPath}: set assignment.owner`);
      continue;
    }

    if (action === "reset_verification_commands") {
      resetVerificationCommands(ticketPath, Array.isArray(params.commands) ? params.commands : []);
      applied.push(`${ticketPath}: reset verification.commands`);
      continue;
    }

    if (action === "normalize_verification_commands") {
      normalizeVerificationCommands(ticketPath);
      applied.push(`${ticketPath}: normalized verification.commands`);
      continue;
    }

    if (action === "set_log_event_type") {
      setLogEventType(logPath);
      applied.push(`${logPath}: set event_type`);
      continue;
    }

    if (action === "normalize_log_context") {
      normalizeLogContext(logPath);
      applied.push(`${logPath}: normalized context`);
      continue;
    }

    if (nonInteractive) {
      throw new Error(`Unsupported repair action ${action}`);
    }
  }

  return applied;
}

export async function runInteractive(repairs, options = {}) {
  const includeOptional = options.includeOptional ?? false;
  if (!process.stdin.isTTY || !process.stdout.isTTY) {
    throw new Error("Interactive mode requires a TTY");
  }

  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  const prepared = [];

  try {
    for (const repair of repairs) {
      if (repair.optional && !includeOptional) {
        continue;
      }

      const [description, suggested] = describeRepair(repair);
      process.stdout.write(`\nRepair ${repair.id}: ${description}\n`);
      const applyChoice = await promptYesNo(rl, "Apply this repair?", true);
      if (!applyChoice) {
        continue;
      }

      repair.enabled = true;
      const action = repair.action;
      const params = repair.params ?? {};

      if (action === "set_front_matter_field") {
        params.value = await promptValueForField(rl, params.field, repair.ticket_path, suggested);
      } else if (action === "set_assignment_owner") {
        params.value = await promptValueForField(rl, "assignment.owner", repair.ticket_path, suggested);
      } else if (action === "reset_verification_commands") {
        params.commands = await promptCommands(rl, Array.isArray(suggested) ? suggested : []);
      }

      repair.params = params;
      prepared.push(repair);
    }
  } finally {
    rl.close();
  }

  return applyRepairs(prepared, { nonInteractive: true, includeOptional });
}

function describeRepair(repair) {
  const action = repair.action;
  const field = repair.params?.field;
  const ticketPath = repair.ticket_path ?? "";
  const logPath = repair.log_path ?? "";
  const value = repair.params?.value;

  if (action === "add_sections") {
    return [`Add missing required sections to ${ticketPath}.`, null];
  }
  if (action === "normalize_created_at") {
    return [`Normalize created_at to ISO8601 UTC in ${ticketPath}.`, iso8601(nowUtc())];
  }
  if (action === "set_front_matter_field") {
    if (field === "id") {
      return ["Set ticket id to a valid UUIDv7 (used to identify the ticket).", newUuidv7()];
    }
    if (field === "version") {
      return [`Set format version (integer, current ${FORMAT_VERSION}).`, FORMAT_VERSION];
    }
    if (field === "version_url") {
      return ["Set version_url (path to the format definition for this version).", FORMAT_VERSION_URL];
    }
    if (field === "priority") {
      return ["Set priority (low|medium|high|critical).", "medium"];
    }
    if (field === "labels") {
      return ["Reset labels to a list of strings (comma-separated).", []];
    }
    return [`Set front matter field '${field}'.`, value];
  }
  if (action === "normalize_labels") {
    return ["Normalize labels to strings, dropping invalid entries.", null];
  }
  if (action === "set_assignment_owner") {
    return ["Set assignment.owner (who owns this ticket; freeform handle).", value];
  }
  if (action === "reset_verification_commands") {
    return ["Set verification.commands (commands to verify acceptance).", value ?? []];
  }
  if (action === "normalize_verification_commands") {
    return ["Normalize verification.commands to strings, dropping invalid entries.", null];
  }
  if (action === "set_log_event_type") {
    return [`Set event_type in ${logPath} based on the log entry shape.`, null];
  }
  if (action === "normalize_log_context") {
    return [`Normalize context in ${logPath} and synthesize a minimal fallback when required.`, null];
  }
  return ["Apply repair", value];
}

async function promptYesNo(rl, message, defaultValue) {
  const suffix = defaultValue ? " [Y/n] " : " [y/N] ";
  while (true) {
    const response = (await rl.question(`${message}${suffix}`)).trim().toLowerCase();
    if (!response) {
      return defaultValue;
    }
    if (["y", "yes"].includes(response)) {
      return true;
    }
    if (["n", "no"].includes(response)) {
      return false;
    }
    process.stdout.write("Please enter y or n.\n");
  }
}

async function promptValueForField(rl, field, ticketPath, defaultValue) {
  let current = null;
  if (ticketPath) {
    try {
      const [frontMatter] = loadTicket(ticketPath);
      current = field === "assignment.owner" ? frontMatter.assignment?.owner : frontMatter[field];
    } catch {
      current = null;
    }
  }

  if (field === "labels") {
    const labels = Array.isArray(defaultValue)
      ? defaultValue
      : Array.isArray(current)
      ? current
      : [];
    return promptLabels(rl, labels);
  }

  if (field === "priority") {
    const base = defaultValue ?? current ?? "medium";
    while (true) {
      const response = (await rl.question(`Priority [${base}]: `)).trim().toLowerCase();
      if (!response) {
        return base;
      }
      if (PRIORITY_VALUES.includes(response)) {
        return response;
      }
      process.stdout.write(`Enter one of ${PRIORITY_VALUES.join(", ")}.\n`);
    }
  }

  const fallback = defaultValue ?? current;
  const response = (await rl.question(`${field} [${fallback ?? ""}]: `)).trim();
  if (!response && fallback != null) {
    return fallback;
  }
  if (field === "id") {
    return response || newUuidv7();
  }
  return response || fallback;
}

async function promptLabels(rl, defaultLabels) {
  const existing = defaultLabels.join(", ");
  const response = (await rl.question(`Labels (comma-separated) [${existing}]: `)).trim();
  if (!response) {
    return defaultLabels;
  }
  return response
    .split(",")
    .map((v) => v.trim())
    .filter(Boolean);
}

async function promptCommands(rl, defaultCommands) {
  const existing = defaultCommands.join("; ");
  process.stdout.write("Enter verification commands (comma-separated). Leave blank to keep default.\n");
  const response = (await rl.question(`Commands [${existing}]: `)).trim();
  if (!response) {
    return defaultCommands;
  }
  return response
    .split(",")
    .map((v) => v.trim())
    .filter(Boolean);
}

function setFrontMatterField(ticketPath, field, value) {
  const [frontMatter, body] = loadTicket(ticketPath);
  frontMatter[field] = value;
  writeTicket(ticketPath, frontMatter, body);
}

function addMissingSections(ticketPath) {
  const [frontMatter, body] = loadTicket(ticketPath);
  let nextBody = body;
  for (const section of ["# Ticket", "## Description", "## Acceptance Criteria", "## Verification"]) {
    if (!nextBody.includes(section)) {
      nextBody += `\n${section}\n(fill in)\n`;
    }
  }
  writeTicket(ticketPath, frontMatter, nextBody);
}

function normalizeCreatedAt(ticketPath) {
  const [frontMatter, body] = loadTicket(ticketPath);
  if (typeof frontMatter.created_at !== "string" || !parseIso(frontMatter.created_at)) {
    frontMatter.created_at = iso8601(nowUtc());
    writeTicket(ticketPath, frontMatter, body);
  }
}

function normalizeLabels(ticketPath) {
  const [frontMatter, body] = loadTicket(ticketPath);
  const labels = Array.isArray(frontMatter.labels)
    ? frontMatter.labels.filter((v) => typeof v === "string")
    : [];
  frontMatter.labels = labels;
  writeTicket(ticketPath, frontMatter, body);
}

function setAssignmentOwner(ticketPath, value) {
  const [frontMatter, body] = loadTicket(ticketPath);
  const assignment =
    frontMatter.assignment && typeof frontMatter.assignment === "object" && !Array.isArray(frontMatter.assignment)
      ? frontMatter.assignment
      : {};
  assignment.owner = value;
  frontMatter.assignment = assignment;
  writeTicket(ticketPath, frontMatter, body);
}

function resetVerificationCommands(ticketPath, commands) {
  const [frontMatter, body] = loadTicket(ticketPath);
  frontMatter.verification = { commands };
  writeTicket(ticketPath, frontMatter, body);
}

function normalizeVerificationCommands(ticketPath) {
  const [frontMatter, body] = loadTicket(ticketPath);
  const rawCommands = Array.isArray(frontMatter.verification?.commands)
    ? frontMatter.verification.commands
    : [];
  frontMatter.verification = {
    commands: rawCommands.filter((v) => typeof v === "string"),
  };
  writeTicket(ticketPath, frontMatter, body);
}

function setLogEventType(logPath) {
  updateJsonl(logPath, (entry) => {
    entry.event_type = inferLogEventType(entry);
    return entry;
  });
}

function normalizeLogContext(logPath) {
  updateJsonl(logPath, (entry) => {
    if (!Array.isArray(entry.context)) {
      entry.context = [];
    }
    entry.context = entry.context.map((value) => String(value).trim()).filter(Boolean);

    const machineEntry = entry.written_by === "tickets" || entry.machine === true;
    const eventType = inferLogEventType(entry);
    if (entry.event_type !== eventType) {
      entry.event_type = eventType;
    }
    if (machineEntry && eventType === "work" && entry.context.length === 0) {
      entry.context = buildFallbackContext(entry);
    }
    return entry;
  });
}

function buildFallbackContext(entry) {
  const items = [];
  if (typeof entry.created_from === "string" && entry.created_from.trim()) {
    items.push(`Handoff from parent ticket ${entry.created_from.trim()}.`);
  }
  if (typeof entry.summary === "string" && entry.summary.trim()) {
    items.push(`Recovered from summary: ${entry.summary.trim()}`);
  }
  if (items.length === 0) {
    items.push("Recovered context was unavailable during automated repair.");
  }
  return items;
}

function inferLogEventType(entry) {
  if (typeof entry.summary === "string" && /^Status (changed|reaffirmed)\b/.test(entry.summary)) {
    return "status";
  }
  return "work";
}

function updateJsonl(logPath, updateEntry) {
  const entries = readJsonl(logPath);
  const content = entries.map((entry) => JSON.stringify(updateEntry(entry))).join("\n");
  fs.writeFileSync(logPath, `${content}\n`);
}
